libpod/image/pull.go

```diff
@@ -114,64 +114,113 @@ func (ir *Runtime) getSinglePullRefPairGoal(srcRef types.ImageReference, destNam
+	refPairs := []pullRefPair{}
+	for _, refList := range refLists {
+		// Note that we are looking only at the first item of the
+		// refList as Podman normalizes short images differently than
+		// c/image.  While c/image is normalizing to docker.io (just as
+		// Docker), Podman is normalizing to "localhost/".  Hence, if
+		// RepoTags are present we normalize them OR we generate the
+		// digest and use that as a destination name.
+		ref := refList[0]
+		destNames, err := reader.ManifestTagsForReference(ref)
 		if err != nil {
 			return nil, err
 		}
-		defer tarSource.Close()
-		manifest, err := tarSource.LoadTarManifest()
 
-		if err != nil {
-			return nil, errors.Wrapf(err, "error retrieving manifest.json")
-		}
-		// to pull the first image stored in the tar file
-		if len(manifest) == 0 {
-			// use the hex of the digest if no manifest is found
-			reference, err := getImageDigest(ctx, srcRef, sc)
+		if len(destNames) == 0 {
+			destName, err := getImageDigest(ctx, ref, sc)
 			if err != nil {
 				return nil, err
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, reference)
+			destNames = append(destNames, destName)
+		} else {
+			for i := range destNames {
+				ref, err := NormalizedTag(destNames[i])
+				if err != nil {
+					return nil, err
+				}
+				destNames[i] = ref.String()
+			}
 		}
 
-		if len(manifest[0].RepoTags) == 0 {
-			// If the input image has no repotags, we need to feed it a dest anyways
-			digest, err := getImageDigest(ctx, srcRef, sc)
+		for _, destName := range destNames {
+			destRef, err := is.Transport.ParseStoreReference(ir.store, destName)
 			if err != nil {
-				return nil, err
+				return nil, errors.Wrapf(err, "error parsing dest reference name %#v", destName)
 			}
-			return ir.getSinglePullRefPairGoal(srcRef, digest)
+			if srcRef != nil {
+				ref = srcRef
+			}
+			pair := pullRefPair{
+				image:  destName,
+				srcRef: ref,
+				dstRef: destRef,
+			}
+			refPairs = append(refPairs, pair)
 		}
+	}
 
-		// Need to load in all the repo tags from the manifest
-		res := []pullRefPair{}
-		for _, dst := range manifest[0].RepoTags {
-			//check if image exists and gives a warning of untagging
-			localImage, err := ir.NewFromLocal(dst)
-			imageID := strings.TrimSuffix(manifest[0].Config, ".json")
-			if err == nil && imageID != localImage.ID() {
-				logrus.Errorf("the image %s already exists, renaming the old one with ID %s to empty string", dst, localImage.ID())
-			}
+	return refPairs, nil
+}
 
-			pullInfo, err := ir.getPullRefPair(srcRef, dst)
-			if err != nil {
-				return nil, err
+// getPullRefPairFromDockerReference returns a slice of pullRefPairs for the specified docker reference.
+func (ir *Runtime) getPullRefPairsFromDockerReference(ctx context.Context, ref types.ImageReference, sc *types.SystemContext) ([]pullRefPair, error) {
+	// Let's first assume it's a file-only reference.
+	reader, err := archive.NewReader(sc, ref.StringWithinTransport())
```

Comment:
No; the caller knows whether the intent is to read everything in a file, or to “pull” a single image. No need to guess.

Reply:
But we don't without breaking backwards compatibility with `podman pull docker-archive:...` which expects all RepoTags to be loaded. Or am I missing something in the c/image PR that I could use?

Reply:
I thought we agreed that using `podman pull` to pull more than one _image_ would be dropped. If it is a single-image archive, `ManifestTagsForReference` lists all tags for that image.

Reply:
Initially yes, but I thought we agreed yesterday to keep extracting all RepoTags to keep the current behaviour.

Reply:
I thought that meant “all tags _for a single image_” — after all, current `podman pull` only ever uses `manifest[0]`. There’s no precedent for `podman pull` actually reading more than one image AFAICS (if only because docker/archive never made that possible).

Reply:
fixed